marc.1
.TH MARC 1 "25 Nov 1992" "Howard Chu@JPL" "LOCAL COMMANDS"
.SH NAME
marc \- archive merger
.SH SYNOPSIS
.B marc
\fItgtarc\fR \fIsrcarc\fR [ \fIfiles\fR ... ]
.SH DESCRIPTION
Reads \fIfiles\fR from the arc archive \fIsrcarc\fR and add them to the
arc archive \fItgtarc\fR.
.SH HISTORY
\fIArc\fP has been in use in the CP/M and MSDOS world for many years.
Thom Henderson developed the original version, but it is important to note that
\fIarc\fP is based on the file compression theories developed by Huffman, Welch,
Knott, Knuth, and many other scientists. This implementation is based on
version 5.21 of the MSDOS program.
.SH BUGS
\fIArc\fP behaves just like the PC version of the program; all functions
of the "usage" display are working.
Full compatibility with PC ARC files is maintained, the price for which is
that \fIarc\fP doesn't like long filenames, and can only archive files with
names of up to 12 characters.
It will *sometimes* do The Right Thing with them, but I suggest
you put long-winded filenames in a "shar" before
.IR arc ing
them.

There shouldn't be any problems, (hah!) but if you find any, please
send them to me at:

	hyc@hanauma.jpl.nasa.gov

.SH AUTHORS
Original MSDOS program by Thom Henderson
.br
COPYRIGHT(C) 1985-87 by System Enhancement Associates; ALL RIGHTS RESERVED

Original Lempel-Zev code derived from compress 4.0.
Modified to support Squashing by Dan Lanciani (ddl@harvard.edu)
Ported from MSDOS by Howard Chu,
with help from John Gilmore (hoptoad!gnu), James Turner (daisy!turner)
and others.


Changelog
Unreleased changes:

Klaus Reimer <k@ailis.de>, 01 Jan 2011:
  * Corrected spelling error in arc man page.
  * Version in binaries now comes from VERSION variable in Makefile.
  * Changed man page sections from 1L to 1.


                        CHANGES IN VERSION 5.21p
                 (Klaus Reimer <k@ailis.de>, 07 Aug 2010)

Replaced mktemp with mkstemp.
Set correct version number and updated creation date in command-line help.
Corrected content of PATCHLEVEL file.
Added man page for marc.
Added install make target.
Added -Wall compiler flag.


                        CHANGES IN VERSION 5.21o
                 (Howard Chu <hyc@highlandsun.com>, 12 Oct 2005)

Silence warnings about unused results.


                        CHANGES IN VERSION 5.21n
                 (Howard Chu <hyc@highlandsun.com>, 08 Oct 2005)

Better fix for temporary file creations, previous patch did not fix marc.
Silence all gcc-4.0.0 -Wall warnings.


                        CHANGES IN VERSION 5.21m
                 (Klaus Reimer <k@ailis.de>, 21 Sep 2005)

Applied patch from Joey Schulze fixing the following problems:
  
  - Fixed insecure permissions of a temporary file [arcsvc.c, CAN-2005-2945]
  - Fixed insecure temporary file creation [arcsvc.c, CAN-2005-2992]



                        CHANGES IN VERSION 5.21l
                 (Klaus Reimer <k@ailis.de>, 13 Mar 2005)
                        
Fixed warnings because missing inclusion of string.h and fixed a gcc-4.0
problem with marc.c (Missing declaration of makefnam).



                        CHANGES IN VERSION 5.21k
                 (Klaus Reimer <k@ailis.de>, 6 Nov 2004)
                        
Ulf Harnhammar found some buffer overflows causing segfaults and provided a
patch which is now applied to arc to close these bugs.



                        CHANGES IN VERSION 5.21


The following changes have been made in ARC version 5.21, but have not
yet been included in the manual:

o   When adding files to an archive, ARC now shows the stowage facter
    achieved on each file.

o   A bug was found that would keep an archive entry from being
    encrypted if it was stored without compression.  This has now been
    fixed.

o   If changes are made to a corrupted archive, the corrupted entries
    are discarded.  This makes it possible to lose data accidentally.
    ARC will now only make changes to a corrupted archive if the W
    (suppress Warnings) option has been given.

o   The N (suppress Notes) option now suppresses the "Creating
    archive" note when a new archive is being created.

o   The N (suppress Notes) option formerly did nothing useful when
    used with the L (List files) command.  It now causes a terse
    listing of filenames only, suitable for use with pipes and
    redirection.

o   The list of filenames given to ARC may now include indirect
    references.  If a filename begins with an "at sign" ("@"), it is
    taken to be the name of a file which contains a list of file
    names.  The list of file names may include further indirection.
    If no extension is given, ".CMD" is assumed.  For example, the
    command:

         arc a waste junk.txt @trash

    would cause ARC to add JUNK.TXT plus all files listed in the file
    TRASH.CMD to an archive named WASTE.ARC.  If no file is specified,
    then the list is read from standard input.  For example, the
    command:

         arc ln waste | arc a trash @

    would cause ARC to add files to TRASH.ARC based on the names of
    the files stored in WASTE.ARC.  It is probably a good idea to give
    the O (Overwrite) option if you are extracting files this way.

    Version 5.21 of MARC also allows for indirection.


marc.c
/*
 * $Header: /cvsroot/arc/arc/marc.c,v 1.7 2011/01/01 13:02:30 k_reimer Exp $
 */

/*  MARC - Archive merge utility

    Version 5.21, created on 04/22/87 at 15:05:10

(C) COPYRIGHT 1985-87 by System Enhancement Associates; ALL RIGHTS RESERVED

    By:	 Thom Henderson

    Description:
	 This program is used to "merge" archives.  That is, to move
	 files from one archive to another with no data conversion.
	 Please refer to the ARC source for a description of archives
	 and archive formats.

    Instructions:
	 Run this program with no arguments for complete instructions.

    Language:
	 Computer Innovations Optimizing C86
*/
#include <stdio.h>
#include <string.h>
#include "arc.h"

#if	UNIX
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#endif

#include "proto.h"

#ifndef	__STDC__
char *calloc(), *malloc(), *realloc(); /* memory managers */
#endif
int gethdr(), match(), readhdr();
VOID	arcdie(), filecopy(), setstamp(), writehdr();
static VOID copyfile(), expandlst(), merge();

FILE *src;			       /* source archive */
char srcname[STRLEN];		       /* source archive name */
char *pinbuf;				/* block input buffer */

static char **lst;		       /* files list */
static int lnum;		       /* length of files list */

int
main(nargs,arg)			       /* system entry point */
int nargs;			       /* number of arguments */
char *arg[];			       /* pointers to arguments */
{
    char *makefnam();
    char *envfind();
#if	!_MTS
    char *arctemp2;		       /* temp file stuff */
#endif
#if	GEMDOS
    VOID exitpause();
#endif
    int n;			       /* index */
#if	UNIX
    struct	stat	sbuf;
#endif


    if(nargs<3)
    {	 printf("MARC - Archive merger, Version %s, created on 08/07/2010\n", VERSION);
/*	 printf("(C) COPYRIGHT 1985,86,87 by System Enhancement Associates;");
	 printf(" ALL RIGHTS RESERVED\n\n");
	 printf("Please refer all inquiries to:\n\n");
	 printf("	System Enhancement Associates\n");
	 printf("	21 New Street, Wayne NJ 07470\n\n");
	 printf("You may copy and distribute this program freely,");
	 printf(" provided that:\n");
	 printf("    1)	  No fee is charged for such copying and");
	 printf(" distribution, and\n");
	 printf("    2)	  It is distributed ONLY in its original,");
	 printf(" unmodified state.\n\n");
	 printf("If you like this program, and find it of use, then your");
	 printf(" contribution will\n");
	 printf("be appreciated.  You may not use this product in a");
	 printf(" commercial environment\n");
	 printf("or a governmental organization without paying a license");
	 printf(" fee of $35.  Site\n");
	 printf("licenses and commercial distribution licenses are");
	 printf(" available.  A program\n");
	 printf("disk and printed documentation are available for $50.\n");
	 printf("\nIf you fail to abide by the terms of this license, ");
	 printf(" then your conscience\n");
	 printf("will haunt you for the rest of your life.\n\n");
*/
	 printf("Usage: MARC <tgtarc> <srcarc> [<filename> . . .]\n");
	 printf("Where: <tgtarc> is the archive to add files to,\n");
	 printf("	<srcarc> is the archive to get files from, and\n");
	 printf("	<filename> is zero or more file names to get.\n");
	 printf("\nAdapted from MSDOS by Howard Chu\n");
#if	GEMDOS
	 exitpause();
#endif
	 return 1;
    }

	/* see where temp files go */
#if	!_MTS
	arctemp = calloc(1, STRLEN);
	if (!(arctemp2 = envfind("ARCTEMP")))
		arctemp2 = envfind("TMPDIR");
	if (arctemp2) {
		strncpy(arctemp, arctemp2, STRLEN - 16);
		arctemp[STRLEN - 17] = '\0';
		n = strlen(arctemp);
		if (arctemp[n - 1] != CUTOFF)
			arctemp[n] = CUTOFF;
	}
#if	UNIX
	else	strcpy(arctemp, "/tmp/");
#endif
#if	!MSDOS
	{
		strcat(arctemp, "AXXXXXX");
		int fd = mkstemp(arctemp);
		if (fd == -1)
		{
			fprintf(stderr, "Unable to create temporary files: %s\n", strerror(errno));
			exit(235);
		}
		close(fd);
		unlink(arctemp);
	}
#else
	strcat(arctemp, "$ARCTEMP");
#endif
#else
	guinfo("SHFSEP	", gotinf);
	sepchr[0] = gotinf[0];
	guinfo("SCRFCHAR", gotinf);
	tmpchr[0] = gotinf[0];
	arctemp = "-$$$";
	arctemp[0] = tmpchr[0];
#endif

	for (n = 1; n < nargs; n++)
	  if (strlen(arg[n]) > (STRLEN - 16))
	  {
	    fprintf(stderr, "Too long argument: %s\n", arg[n]);
	    exit(235);
	  }

#if	UNIX
	if (!stat(arg[1],&sbuf))
		strcpy(arcname,arg[1]);
	else
		makefnam(arg[1],".arc",arcname);
	if (!stat(arg[2],&sbuf))
		strcpy(srcname,arg[2]);
	else
		makefnam(arg[2],".arc",srcname);
#else
    makefnam(arg[1],".ARC",arcname);   /* fix up archive names */
    makefnam(arg[2],".ARC",srcname);
/*	makefnam(".$$$",arcname,newname);*/
#endif
	sprintf(newname,"%s.arc",arctemp);

    arc = fopen(arcname,OPEN_R);	       /* open the archives */
    if(!(src=fopen(srcname,OPEN_R)))
	 arcdie("Cannot read source archive %s",srcname);
    if(!(new=tmpopen(newname)))
	 arcdie("Cannot create new archive %s",newname);

    if(!arc)
	 printf("Creating new archive %s\n",arcname);

    /* get the files list set up */

    lnum = nargs-3;		       /* initial length of list */
    if(lnum<1)			       /* phoney for default case */
    {	 lnum = 1;
	 lst = (char **) calloc(1,sizeof(char *));
	 lst[0] = "*.*";
    }
    else			       /* else use filenames given */
    {	 lst = (char **) calloc(lnum,sizeof(char *));
	 for(n=3; n<nargs; n++)
	      lst[n-3] = arg[n];

	 for(n=0; n<lnum; )	       /* expand indirect references */
	 {    if(*lst[n] == '@')
		   expandlst(n);
	      else n++;
	 }
    }
    if (!(pinbuf = malloc(MYBUF)))
	arcdie("Not enough memory for input buffer.");

    merge(lnum,lst);		       /* merge desired files */

    if(arc) fclose(arc);	       /* close the archives */
    fclose(src);
    fclose(new);

    if(arc)			       /* make the switch */
	 if(unlink(arcname))
	      arcdie("Unable to delete old copy of %s",arcname);
    if(move(newname,arcname))
	 arcdie("Unable to rename %s to %s",newname,arcname);

    setstamp(arcname,arcdate,arctime);     /* new arc matches newest file */

#if	GEMDOS
    exitpause();
#endif
    return nerrs;
}

static	VOID
merge(nargs,arg)		       /* merge two archives */
int nargs;			       /* number of filename templates */
char *arg[];			       /* pointers to names */
{
    struct heads srch;		       /* source archive header */
    struct heads arch;		       /* target archive header */
    int gotsrc, gotarc;		       /* archive entry versions (0=end) */
    int copy;			       /* true to copy file from source */
    int n;			       /* index */

    gotsrc = gethdr(src,&srch);	       /* get first source file */
    gotarc = gethdr(arc,&arch);	       /* get first target file */

    while(gotsrc || gotarc)	       /* while more to merge */
    {	 if(strcmp(srch.name,arch.name)>0)
	 {    copyfile(arc,&arch,gotarc);
	      gotarc = gethdr(arc,&arch);
	 }

	 else if(strcmp(srch.name,arch.name)<0)
	 {    copy = 0;
	      for(n=0; n<nargs; n++)
	      {	   if(match(srch.name,arg[n]))
		   {	copy = 1;
			break;
		   }
	      }
	      if(copy)		       /* select source or target */
	      {	   printf("Adding file:	  %s\n",srch.name);
		   copyfile(src,&srch,gotsrc);
	      }
	      else fseek(src,srch.size,1);
	      gotsrc = gethdr(src,&srch);
	 }

	 else			       /* duplicate names */
	 {    copy = 0;
	      {	   if((srch.date>arch.date)
		   || (srch.date==arch.date && srch.time>arch.time))
		   {	for(n=0; n<nargs; n++)
			{    if(match(srch.name,arg[n]))
			     {	  copy = 1;
				  break;
			     }
			}
		   }
	      }
	      if(copy)		       /* select source or target */
	      {	   printf("Updating file: %s\n",srch.name);
		   copyfile(src,&srch,gotsrc);
		   gotsrc = gethdr(src,&srch);
		   if(gotarc)
		   {	fseek(arc,arch.size,1);
			gotarc = gethdr(arc,&arch);
		   }
	      }
	      else
	      {	   copyfile(arc,&arch,gotarc);
		   gotarc = gethdr(arc,&arch);
		   if(gotsrc)
		   {	fseek(src,srch.size,1);
			gotsrc = gethdr(src,&srch);
		   }
	      }
	 }
    }

    hdrver = 0;			       /* end of archive marker */
    writehdr(&arch,new);	       /* mark the end of the archive */
}

int gethdr(f,hdr)		       /* special read header for merge */
FILE *f;			       /* file to read from */
struct heads *hdr;		       /* storage for header */
{
    char *i = hdr->name;	       /* string index */
    int n;			       /* index */

    for(n=0; n<FNLEN; n++)	       /* fill name field */
	 *i++ = 0176;		       /* impossible high value */
    *--i = '\0';		       /* properly end the name */

    hdrver = 0;			       /* reset header version */
    if(readhdr(hdr,f))		       /* use normal reading logic */
	 return hdrver;		       /* return the version */
    else return 0;		       /* or fake end of archive */
}

static VOID
copyfile(f,hdr,ver)		       /* copy a file from an archive */
FILE *f;			       /* archive to copy from */
struct heads *hdr;		       /* header data for file */
int ver;			       /* header version */
{
    hdrver = ver;		       /* set header version */
    writehdr(hdr,new);		       /* write out the header */
    filecopy(f,new,hdr->size);	       /* copy over the data */
}

static VOID
expandlst(n)			       /* expand an indirect reference */
int n;				       /* number of entry to expand */
{
    FILE *lf, *fopen();		       /* list file, opener */
    char buf[100];		       /* input buffer */
    int x;			       /* index */
    char *p = lst[n]+1;		       /* filename pointer */
    char *makefnam();

    if(*p)			       /* use name if one was given */
    {	 makefnam(p,".CMD",buf);
	 upper(buf);
	 if(!(lf=fopen(buf,"r")))
	      arcdie("Cannot read list of files in %s",buf);
    }
    else lf = stdin;		       /* else use standard input */

    for(x=n+1; x<lnum; x++)	       /* drop reference from the list */
	 lst[x-1] = lst[x];
    lnum--;

    while(fscanf(lf,"%99s",buf)>0)     /* read in the list */
    {	 if(!(lst=(char **) realloc(lst,(lnum+1)*sizeof(char *))))
	      arcdie("too many file references");

	 lst[lnum] = malloc(strlen(buf)+1);
	 strcpy(lst[lnum],buf);	       /* save the name */
	 lnum++;
    }

    if(lf!=stdin)		       /* avoid closing standard input */
	 fclose(lf);
}
